Hook up the new frameworks with Actionable Error (#36455)
rails/rails: activesupport/CHANGELOG.md

*   Trigger `ActiveSupport::ActionableError` from existing errors.

>     Triggers let's you re raise an actionable error from an existing one, if it

Reviewer: `re-raise` since it should be one word.

Author: Thanks! Fixed it.

*   Introduce `ActiveSupport::ActionableError.define`.

>     This can be useful if you're code does not depend strictly on Rails. In

Reviewer 2: `your`. Should this say Rails or just Active Support? It's a weird boundary that we mention Rails in the second line even though this is an Active Support feature. So maybe we should mention how to catch and triggers these in people's own apps?

Reviewer 2: Riiiight, we're trying to allude to that this depends on the Action Pack raise/rescue to get built in support.

    this case, you may not want to define an error depending on Rails specific
    code next to your regular error class hierarchy.

>     You can define it the actionable error straight in a Railtie, next to you

Reviewer 2: ```suggestion
    You can define the actionable error in a Railtie, next to your
```

    Rails dependent code:

    ```ruby
    class ActiveStorage::Railtie < Rails::Engine
      initializer "active_storage.actionable_errors" do
        ActiveSupport::ActionableError.define :InstallError, under: ActiveStorage do |actionable|
          actionable.message <<~END
            Action Mailbox does not appear to be installed. Do you want to install it now?
          END

          actionable.trigger on: ActiveRecord::StatementInvalid, if: -> error do
            error.message.match?(InboundEmail.table_name)
          end

          actionable.action "Install now" do
            Rails::Command.invoke("active_storage:install")
            Rails::Command.invoke("db:migrate")
          end
        end
      end
    end
    ```

    *Genadi Samokovarov*

*   Trigger `ActiveSupport::ActionableError` from existing errors.

>     Triggers let's you re raise an actionable error from an existing one, if it

Reviewer 2: ```suggestion
    Triggers lets you re-raise an actionable error from an existing one, if it
```

Lets without apostrophe because let's expands to "let us".

    matches a certain condition. Here is how Action Mailbox uses triggers to
    raise an actionable error when the `action_mailbox_inbound_emails` table
    does not exist.

    ```ruby
    class InstallError < Error
      include ActiveSupport::ActionableError

      def initialize(message = nil)

>         super(message || <<~MESSAGE)

Reviewer 2: I still don't understand why we do `message ||`. Why do we need the default?

          Action Mailbox does not appear to be installed. Do you want to
          install it now?
        MESSAGE
      end

      trigger on: ActiveRecord::StatementInvalid, if: -> error do
        error.message.match?(InboundEmail.table_name)
      end

      action "Install now" do
        Rails::Command.invoke("active_storage:install")

>         Rails::Command.invoke("db:migrate")j

Reviewer 3: The `j` here is probably a typo?

Author: It is, thanks for catching!
